src/Sarif/Visitors/SarifCurrentToVersionOneVisitor.cs

```diff
@@ -904,6 +903,105 @@ internal RunVersionOne CreateRun(Run v2Run)
+
+                for (int index = 0; index < v2Files.Count; ++index)
+                {
+                    FileData v2File = v2Files[index];
+                    string key = CreateFileDictionaryKey(v2File, v2Files);
+
+                    fileKeyToIndexDictionary[key] = index;
+                    fileIndexToKeyDictionary[index] = key;
+                }
+            }
+        }
+
+        // Given a v2 FileData object, synthesize a key that will be used to locate that object
+        // in the v2 file data dictionary. We will use the same key to locate the corresponding
+        // v1 FileData object when we create the v1 run.files dictionary.
+        //
+        // This method is necessary because, although ideally we would use the FileData
+        // object's URI as the dictionary key, it is possible for two FileData objects to
+        // have the same URI. For example:
+        //
+        //    1. Nested files with the same name might exist in two different containers.
+        //    2. There might be two files with the same URI but difference uriBaseIds.
+        //
+        // To deal with Case #1, we synthesize a key as follows:
+        //
+        //    <root file URI>#<nested file path>/<level 2 nested file path>/...
+        //
+        // To deal with Case #2, we would follow the chain or uriBaseIds, prepending
+        // the value of each one to the URI.
+        //
+        // WE DO NOT YET HANDLE CASE #2.
+        private static string CreateFileDictionaryKey(FileData v2File, IList<FileData> v2Files)
+        {
+            var sb = new StringBuilder(v2File.FileLocation.Uri.OriginalString);
+            while (v2File.ParentIndex != -1)
+            {
+                FileData parentFile = v2Files[v2File.ParentIndex];
+
+                // The convention for building the key is as follows:
+                // The root file URI is separated from the chain of nested files by '#';
+                // The nested file URIs are separated from each other by '/'.
+                string separator = parentFile.ParentIndex == -1 ? "#" : "/";
+
+                sb.Insert(0, separator);
+                sb.Insert(0, parentFile.FileLocation.Uri.OriginalString);
+
+                v2File = parentFile;
+            }
+
+            return sb.ToString();
+        }
+
+        private IDictionary<string, FileDataVersionOne> CreateV1FileDictionary()
+        {
+            Dictionary<string, FileDataVersionOne> filesVersionOne = null;
+
+            if (_fileKeyToIndexDictionary != null)
+            {
+                filesVersionOne = new Dictionary<string, FileDataVersionOne>();
+                foreach (KeyValuePair<string, int> entry in _fileKeyToIndexDictionary)
+                {
+                    string key = entry.Key;
+                    int index = entry.Value;
+                    FileData v2File = _currentV2Run.Files[index];
+                    FileDataVersionOne fileDataVersionOne = CreateFileData(v2File);
```

Comment:
>CreateFileData [](start = 60, length = 14)

Again prefer to version these things, e.g., CreateFileDataVersionOne

Reply:
Yes, I agree: I always found this code confusing for that reason, and _all_ the `CreateXxx` methods follow this bad pattern. I'll change them all.

---
In reply to: [247376992](https://github.com/Microsoft/sarif-sdk/pull/1211#discussion_r247376992) [](ancestors = 247376992)

Reply:
Fixed.

---
In reply to: [247565505](https://github.com/Microsoft/sarif-sdk/pull/1211#discussion_r247565505) [](ancestors = 247565505,247376992)